File: libs/idun_agent_standalone/src/idun_agent_standalone/theme/runtime_config.py
"""``GET /runtime-config.js`` — backend-templated bootstrap script.

The static UI is a Next.js export and cannot embed backend-dependent values
at build time. The browser loads this script before the SPA's main bundle
to populate ``window.__IDUN_CONFIG__`` with the active theme and auth mode.
"""

from __future__ import annotations

import json
from typing import Any

from fastapi import APIRouter, Request, Response
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel
from sqlalchemy import select

from idun_agent_standalone.db.models import ThemeRow

router = APIRouter(tags=["runtime-config"])


class _CamelModel(BaseModel):
    """Pydantic base that serializes to camelCase (frontend convention)."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )


class ThemeColors(_CamelModel):
    background: str
    foreground: str
    card: str
    card_foreground: str
    popover: str
    popover_foreground: str
    primary: str
    primary_foreground: str
    secondary: str
    secondary_foreground: str
    muted: str
    muted_foreground: str
    accent: str
    accent_foreground: str
    destructive: str
    destructive_foreground: str
    border: str
    input: str
    ring: str


class LightDarkColors(_CamelModel):
    light: ThemeColors
    dark: ThemeColors


class ThemeLogo(_CamelModel):
    text: str
    image_url: str | None = None


class ThemeConfig(_CamelModel):
    app_name: str
    greeting: str
    starter_prompts: list[str]
    logo: ThemeLogo
    layout: str
    colors: LightDarkColors
    radius: str
    font_sans: str
    font_serif: str
    font_mono: str
    default_color_scheme: str


_EDITORIAL_LIGHT = ThemeColors(
    background="#f7f6f0",
    foreground="#1d1c1a",
    card="#ffffff",
    card_foreground="#1d1c1a",
    popover="#ffffff",
    popover_foreground="#1d1c1a",
    primary="#1d1c1a",
    primary_foreground="#f7f6f0",
    secondary="#f0eee2",
    secondary_foreground="#1d1c1a",
    muted="#f0eee2",
    muted_foreground="#6b6a65",
    accent="#c96442",
    accent_foreground="#ffffff",
    destructive="#dc2626",
    destructive_foreground="#ffffff",
    border="#e7e4d7",
    input="#e7e4d7",
    ring="rgba(201, 100, 66, 0.4)",
)


_EDITORIAL_DARK = ThemeColors(
    background="#15140f",
    foreground="#f5f4ec",
    card="#1d1c1a",
    card_foreground="#f5f4ec",
    popover="#1d1c1a",
    popover_foreground="#f5f4ec",
    primary="#f5f4ec",
    primary_foreground="#15140f",
    secondary="#2a2925",
    secondary_foreground="#f5f4ec",
    muted="#2a2925",
    muted_foreground="#a1a097",
    accent="#d97757",
    accent_foreground="#15140f",
    destructive="#ef4444",
    destructive_foreground="#f5f4ec",
    border="#2a2925",
    input="#2a2925",
    ring="rgba(217, 119, 87, 0.5)",
)


DEFAULT_THEME_MODEL = ThemeConfig(
    app_name="My Assistant",
    greeting="How can I help?",
    starter_prompts=[],
    logo=ThemeLogo(text="MA"),
    layout="branded",
    colors=LightDarkColors(light=_EDITORIAL_LIGHT, dark=_EDITORIAL_DARK),
    radius="0.625",
    font_sans="",
    font_serif="",
    font_mono="",
    default_color_scheme="system",
)


# ``DEFAULT_THEME`` is the camelCase JSON representation of the editorial
# default theme; ``window.__IDUN_CONFIG__.theme`` consumers expect this exact
# shape. Materializing it once at import time keeps the request path cheap.
DEFAULT_THEME: dict[str, Any] = DEFAULT_THEME_MODEL.model_dump(by_alias=True)


@router.get("/runtime-config.js")
async def runtime_config_js(request: Request) -> Response:
    sm = request.app.state.sessionmaker
    settings = request.app.state.settings
    async with sm() as s:
        row = (await s.execute(select(ThemeRow))).scalar_one_or_none()
    theme: dict[str, Any] = {
        **DEFAULT_THEME,
        **((row.config or {}) if row else {}),
    }

    config: dict[str, Any] = {
        "theme": theme,
        "authMode": settings.auth_mode.value,
        "layout": theme.get("layout", "branded"),
    }
    body = f"window.__IDUN_CONFIG__ = {json.dumps(config)};\n"
    return Response(
        content=body,
        media_type="application/javascript",
        headers={"Cache-Control": "no-store"},
    )

File: services/idun_agent_standalone_ui/lib/runtime-config.ts
/**
 * Runtime config injected by the FastAPI backend via /runtime-config.js.
 *
 * The script tag in app/layout.tsx loads /runtime-config.js with
 * strategy="beforeInteractive" so window.__IDUN_CONFIG__ is populated
 * before React hydration. Defaults below mirror the backend so SSR-out
 * static export and dev mode (no backend) still render reasonably.
 */

export type ThemeColors = {
  background: string;
  foreground: string;
  card: string;
  cardForeground: string;
  popover: string;
  popoverForeground: string;
  primary: string;
  primaryForeground: string;
  secondary: string;
  secondaryForeground: string;
  muted: string;
  mutedForeground: string;
  accent: string;
  accentForeground: string;
  destructive: string;
  destructiveForeground: string;
  border: string;
  input: string;
  ring: string;
};

export type ThemeConfig = {
  appName: string;
  greeting: string;
  starterPrompts: string[];
  logo: { text: string; imageUrl?: string };
  layout: "branded" | "minimal" | "inspector";
  colors: { light: ThemeColors; dark: ThemeColors };
  radius: string;
  fontSans: string;
  fontSerif: string;
  fontMono: string;
  defaultColorScheme: "light" | "dark" | "system";
};

export type RuntimeConfig = {
  theme: ThemeConfig;
  authMode: "none" | "password" | "oidc";
  layout: "branded" | "minimal" | "inspector";
};

declare global {
  interface Window {
    __IDUN_CONFIG__?: RuntimeConfig;
  }
}

const EDITORIAL_LIGHT: ThemeColors = {
  background: "#f7f6f0",
  foreground: "#1d1c1a",
  card: "#ffffff",
  cardForeground: "#1d1c1a",
  popover: "#ffffff",
  popoverForeground: "#1d1c1a",
  primary: "#1d1c1a",
  primaryForeground: "#f7f6f0",
  secondary: "#f0eee2",
  secondaryForeground: "#1d1c1a",
  muted: "#f0eee2",
  mutedForeground: "#6b6a65",
  accent: "#c96442",
  accentForeground: "#ffffff",
  destructive: "#dc2626",
  destructiveForeground: "#ffffff",
  border: "#e7e4d7",
  input: "#e7e4d7",
  ring: "rgba(201, 100, 66, 0.4)",
};

const EDITORIAL_DARK: ThemeColors = {
  background: "#15140f",
  foreground: "#f5f4ec",
  card: "#1d1c1a",
  cardForeground: "#f5f4ec",
  popover: "#1d1c1a",
  popoverForeground: "#f5f4ec",
  primary: "#f5f4ec",
  primaryForeground: "#15140f",
  secondary: "#2a2925",
  secondaryForeground: "#f5f4ec",
  muted: "#2a2925",
  mutedForeground: "#a1a097",
  accent: "#d97757",
  accentForeground: "#15140f",
  destructive: "#ef4444",
  destructiveForeground: "#f5f4ec",
  border: "#2a2925",
  input: "#2a2925",
  ring: "rgba(217, 119, 87, 0.5)",
};

const DEFAULT_THEME: ThemeConfig = {
  appName: "Idun Agent",
  greeting: "How can I help?",
  starterPrompts: [],
  logo: { text: "IA" },
  layout: "branded",
  colors: {
    light: EDITORIAL_LIGHT,
    dark: EDITORIAL_DARK,
  },
  radius: "0.625",
  fontSans: "",
  fontSerif: "",
  fontMono: "",
  defaultColorScheme: "system",
};

export const DEFAULT_RUNTIME_CONFIG: RuntimeConfig = {
  theme: DEFAULT_THEME,
  authMode: "none",
  layout: "branded",
};

export function getRuntimeConfig(): RuntimeConfig {
  if (typeof window === "undefined" || !window.__IDUN_CONFIG__) {
    return DEFAULT_RUNTIME_CONFIG;
  }
  return window.__IDUN_CONFIG__;
}
